pkg/distributor/distributor.go

```diff
@@ -1093,11 +1089,56 @@ func (d *Distributor) Push(ctx context.Context, req *mimirpb.WriteRequest) (*mim
 		mimirpb.ReuseSlice(req.Timeseries)
 	})
 
-	err := d.PushWithMiddlewares(ctx, pushReq)
-	if err != nil {
-		return nil, err
+	pushErr := d.PushWithMiddlewares(ctx, pushReq)
+	if pushErr == nil {
+		return &mimirpb.WriteResponse{}, nil
 	}
-	return &mimirpb.WriteResponse{}, nil
+	handledErr := d.handlePushError(ctx, pushErr, nil)
+	return nil, handledErr
+}
+
+func (d *Distributor) handlePushError(ctx context.Context, err error, limits *validation.Overrides) error {
+	if errors.Is(err, context.Canceled) {
+		return err
+	}
+
+	var (
+		replicasNotMatchErr distributorerror.ReplicasNotMatch
+		tooManyClusterErr   distributorerror.TooManyClusters
+		validationErr       distributorerror.Validation
+		ingestionRateErr    distributorerror.IngestionRateLimited
+		requestRateErr      distributorerror.RequestRateLimited
+	)
+
+	switch {
+	case errors.As(err, &replicasNotMatchErr):
+		return httpgrpc.Errorf(http.StatusAccepted, err.Error())
+	case errors.As(err, &tooManyClusterErr):
+		return httpgrpc.Errorf(http.StatusTooManyRequests, err.Error())
```

Comment:
Wasn't returning `StatusBadRequest` before? Now we're returning 429 which may be retried by the client, but this error shouldn't be retried because the retry will get the same error again.